File: Nexus/Source/DefinitionsTests/ExchangeRateTester.cpp
#include <Beam/SerializationTests/ValueShuttleTests.hpp>
#include <Beam/Utilities/ToString.hpp>
#include <doctest/doctest.h>
#include "Nexus/Definitions/ExchangeRate.hpp"

using namespace Beam;
using namespace Beam::Tests;
using namespace boost;
using namespace Nexus;
using namespace Nexus::DefaultCurrencies;

TEST_SUITE("ExchangeRate") {
  TEST_CASE("default_constructor") {
    auto rate = ExchangeRate();
    REQUIRE(rate.m_rate.numerator() == 1);
    REQUIRE(rate.m_rate.denominator() == 1);
    REQUIRE(rate.m_pair == CurrencyPair());
  }

  TEST_CASE("constructor") {
    auto pair = CurrencyPair(USD, EUR);
    auto rate  = rational<int>(3, 4);
    auto exchange_rate = ExchangeRate(pair, rate);
    REQUIRE(exchange_rate.m_rate == rate);
    REQUIRE(exchange_rate.m_pair == pair);
  }

  TEST_CASE("invert") {
    auto pair = CurrencyPair(GBP, JPY);
    auto rate  = rational<int>(2, 5);
    auto exchange_rate = ExchangeRate(pair, rate);
    auto inverse_rate = invert(exchange_rate);
    REQUIRE(inverse_rate.m_rate.numerator() == rate.denominator());
    REQUIRE(inverse_rate.m_rate.denominator() == rate.numerator());
    REQUIRE(inverse_rate.m_pair == invert(pair));
  }

  TEST_CASE("invert_twice") {
    auto pair = CurrencyPair(AUD, CAD);
    auto rate  = rational<int>(7, 9);
    auto exchange_rate = ExchangeRate(pair, rate);
    auto back = invert(invert(exchange_rate));
    REQUIRE(back == exchange_rate);
  }

  TEST_CASE("convert") {
    auto rate = rational<int>(3, 2);
    auto original = 100 * Money::ONE;
    auto exchange_rate = ExchangeRate(CurrencyPair(EUR, GBP), rate);
    auto converted = convert(original, exchange_rate);
    auto expected  = (rate.numerator() * original) / rate.denominator();
    REQUIRE(converted == expected);
  }

  TEST_CASE("convert_then_invert") {
    auto rate = rational<int>(5, 3);
    auto original = 45 * Money::ONE;
    auto exchange_rate = ExchangeRate(CurrencyPair(USD, CAD), rate);
    auto converted = convert(original, exchange_rate);
    auto back = convert(converted, invert(exchange_rate));
    REQUIRE(back == original);
  }

  TEST_CASE("stream") {
    auto pair = CurrencyPair(EUR, USD);
    auto rate = ExchangeRate(pair, rational<int>(3, 2));
    CHECK(to_string(rate) == "(EUR/USD 3/2)");
    test_round_trip_shuttle(rate);
  }
}

File: WebApi/tests/definitions/exchange_rate_tester.ts
import { describe, it } from 'node:test';
import * as assert from 'node:assert';
import Fraction from 'fraction.js';
import { Currency } from '../../source/definitions/currency';
import { CurrencyPair } from '../../source/definitions/currency_pair';
import { ExchangeRate } from '../../source/definitions/exchange_rate';
import { Money } from '../../source/definitions/money';

describe('ExchangeRate', () => {
  it('construct', () => {
    const pair = new CurrencyPair(new Currency(840), new Currency(978));
    const rate = new Fraction(3, 4);
    const exchangeRate = new ExchangeRate(pair, rate);
    assert.ok(exchangeRate.pair.equals(pair));
    assert.ok(exchangeRate.rate.equals(rate));
  });

  it('none', () => {
    assert.ok(ExchangeRate.NONE.pair.equals(CurrencyPair.NONE));
    assert.ok(ExchangeRate.NONE.rate.equals(new Fraction(1)));
  });

  it('equals', () => {
    const a = new ExchangeRate(
      new CurrencyPair(new Currency(840), new Currency(978)),
      new Fraction(3, 4));
    const b = new ExchangeRate(
      new CurrencyPair(new Currency(840), new Currency(978)),
      new Fraction(3, 4));
    assert.ok(a.equals(b));
  });

  it('not_equal_pair', () => {
    const a = new ExchangeRate(
      new CurrencyPair(new Currency(840), new Currency(978)),
      new Fraction(3, 4));
    const b = new ExchangeRate(
      new CurrencyPair(new Currency(124), new Currency(978)),
      new Fraction(3, 4));
    assert.ok(!a.equals(b));
  });

  it('not_equal_rate', () => {
    const a = new ExchangeRate(
      new CurrencyPair(new Currency(840), new Currency(978)),
      new Fraction(3, 4));
    const b = new ExchangeRate(
      new CurrencyPair(new Currency(840), new Currency(978)),
      new Fraction(5, 4));
    assert.ok(!a.equals(b));
  });

  it('invert', () => {
    const pair = new CurrencyPair(new Currency(826), new Currency(392));
    const rate = new Fraction(2, 5);
    const exchangeRate = new ExchangeRate(pair, rate);
    const inverted = exchangeRate.invert();
    assert.ok(inverted.rate.equals(new Fraction(5, 2)));
    assert.ok(inverted.pair.equals(pair.invert()));
  });

  it('invert_twice', () => {
    const pair = new CurrencyPair(new Currency(36), new Currency(124));
    const rate = new Fraction(7, 9);
    const exchangeRate = new ExchangeRate(pair, rate);
    const back = exchangeRate.invert().invert();
    assert.ok(back.equals(exchangeRate));
  });

  it('convert', () => {
    const rate = new Fraction(3, 2);
    const original = Money.ONE.multiply(100);
    const exchangeRate = new ExchangeRate(
      new CurrencyPair(new Currency(978), new Currency(826)), rate);
    const converted = exchangeRate.convert(original);
    const expected = original.multiply(3).divide(2);
    assert.ok(converted.equals(expected));
  });

  it('convert_then_invert', () => {
    const rate = new Fraction(5, 3);
    const original = Money.ONE.multiply(45);
    const exchangeRate = new ExchangeRate(
      new CurrencyPair(new Currency(840), new Currency(124)), rate);
    const converted = exchangeRate.convert(original);
    const back = exchangeRate.invert().convert(converted);
    assert.ok(back.equals(original));
  });

  it('to_json', () => {
    const exchangeRate = new ExchangeRate(
      new CurrencyPair(new Currency(978), new Currency(840)),
      new Fraction(3, 2));
    const json = exchangeRate.toJson();
    assert.strictEqual(json.pair.base, 978);
    assert.strictEqual(json.pair.counter, 840);
    assert.strictEqual(json.rate.numerator, 3);
    assert.strictEqual(json.rate.denominator, 2);
  });

  it('from_json', () => {
    const json = {
      pair: { base: 978, counter: 840 },
      rate: { numerator: 3, denominator: 2 }
    };
    const exchangeRate = ExchangeRate.fromJson(json);
    assert.ok(exchangeRate.pair.base.equals(new Currency(978)));
    assert.ok(exchangeRate.pair.counter.equals(new Currency(840)));
    assert.ok(exchangeRate.rate.equals(new Fraction(3, 2)));
  });

  it('round_trip_json', () => {
    const original = new ExchangeRate(
      new CurrencyPair(new Currency(840), new Currency(124)),
      new Fraction(7, 4));
    const restored = ExchangeRate.fromJson(original.toJson());
    assert.ok(original.equals(restored));
  });
});
